### PR TITLE
Support stamp specific redirects with Azure exporters

### PR DESCRIPTION
**Pull request by lzchen**
Breeze endpoint returns 307, 308 on redirect response.
Handle redirect in exporter by reading new endpoint location from header, automatically retrying and setting endpoint to new endpoint.

### CODE REVIEW

File: contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py [hunk 1]
```diff
@@ -22,13 +22,22 @@
...
 
+try:
+    from urllib.parse import urlparse
+except ImportError:
+    from urlparse import urlparse
+
+
 logger = logging.getLogger(__name__)
+
+_MAX_CONSECUTIVE_REDIRECTS = 30
```
Comment (hectorhdzg) [inline contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py:33]: This should be 10 according to spec

File: contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py [hunk 2]
```diff
@@ -79,6 +85,7 @@ def _transmit(self, envelopes):
                 headers=headers,
                 timeout=self.options.timeout,
                 proxies=json.loads(self.options.proxies),
+                allow_redirects=False,
```
Comment (hectorhdzg) [inline contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py:88]: Is requests library already handling redirects?, any reason to not just use that?

File: contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py [hunk 3]
```diff
@@ -209,6 +217,25 @@ def _transmit(self, envelopes):
                 with _requests_lock:
                     _requests_map['retry'] = _requests_map.get('retry', 0) + 1  # noqa: E501
             return self.options.minimum_retry_interval
+        if response.status_code in (307, 308):
```
Comment (hectorhdzg) [inline contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py:220]: For 307 case, Cache-Control header would be returned as well specifying how much time the location URL need to be cached

File: contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py (hunk 2 above)
Comment (lzchen) [reply contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py:88]: We need to change the endpoint to the new location and `requests` library does not supply that if we let it handle redirects.

File: contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py (hunk 3 above)
Comment (lzchen) [reply contrib/opencensus-ext-azure/opencensus/ext/azure/common/transport.py:221]: Will probably leave this for another PR.

File: contrib/opencensus-ext-azure/tests/test_transport_mixin.py [hunk 4]
```diff
@@ -47,24 +48,25 @@ def func(*_args, **_kwargs):
...
-        mixin.options = Options()
-        mixin.options.enable_stats_metrics = True
-        self.assertTrue(mixin._check_stats_collection())
-        mixin._is_stats = False
-        self.assertTrue(mixin._check_stats_collection())
-        mixin._is_stats = True
-        self.assertFalse(mixin._check_stats_collection())
-        mixin.options.enable_stats_metrics = False
-        self.assertFalse(mixin._check_stats_collection())
+    # def test_check_stats_collection(self):
```
Comment (hectorhdzg) [inline contrib/opencensus-ext-azure/tests/test_transport_mixin.py:59]: Any reason to disable this one?

File: contrib/opencensus-ext-azure/tests/test_transport_mixin.py [hunk 5]
```diff
@@ -318,3 +321,18 @@ def test_transmission_400(self):
                 post.return_value = MockResponse(400, '{}')
                 mixin._transmit_from_storage()
             self.assertEqual(len(os.listdir(mixin.storage.path)), 0)
+
+    def test_transmission_307(self):
```
Comment (hectorhdzg) [inline contrib/opencensus-ext-azure/tests/test_transport_mixin.py:325]: Can you add a test for _MAX_CONSECUTIVE_REDIRECTS to validate circular reference is handled?

File: contrib/opencensus-ext-azure/tests/test_transport_mixin.py (hunk 4 above)
Comment (lzchen) [reply contrib/opencensus-ext-azure/tests/test_transport_mixin.py:59]: Accident. Re-added.

File: contrib/opencensus-ext-azure/tests/test_transport_mixin.py (hunk 5 above)
Comment (lzchen) [reply contrib/opencensus-ext-azure/tests/test_transport_mixin.py:325]: Added.

**Review by hectorhdzg — APPROVED** [review]

**Pull request merged**